### ISSUE TITLE
Using Example 4: basic cap no callback why is the Local Host data Ethernet II and not IP

### ISSUE THREAD

**Issue by Terry550** [chmorgan/sharppcap#265]
I ran your example 4, basic cap no callback to capture Local Host data, and then opened my saved .pcap file in Wireshark (v3.4.4).
It appears the local host data was recorded but not sure why it shows up in Wireshark as Ethernet II and not IP TCP.
Is this as expected?
Thanks
Terry

**Comment by Terry550** [comment]
I've attached my VS2015 dotnetframework v4.6.1 solution and a pcap file I captured for loopback adapter.
Any help would be greatly appreciated.
Thanks
Terry
[LoopBack_Capture_VS2015_NetFW4.6.1_SharpPcapv5.4.zip](https://github.com/chmorgan/sharppcap/files/6239035/LoopBack_Capture_VS2015_NetFW4.6.1_SharpPcapv5.4.zip)
[EMS_EthernetMonitorServer_WPCAP_NPCAP LocalHost Example.zip](https://github.com/chmorgan/sharppcap/files/6239036/EMS_EthernetMonitorServer_WPCAP_NPCAP.LocalHost.Example.zip)

**Comment by chmorgan** [comment]
Hi @Terry550. What adapter are you capturing from? I'm trying to recall the specifics, something about when you capture at the loopback layer where you end up losing the link layer type comes to mind.

Looking at the captured data it doesn't look like Wireshark can decode any of the packets. What happens if you capture from the same device from Wireshark?

**Comment by Terry550** [comment]
In wire shark the same interface shows up as 127.0.0.1 and Tcp traffic.
I’m using packet dot net version 1.2 is there a newer version than that I
should be using ?thanks for your help.
Terry

**Comment by Terry550** [comment]
Hi chmorgan,
I have more info to provide. I updated to VS2019 and tried the framework 4.8 but had issues compiling it kept saying I need DotNetStandard 2.0 ref but I already had installed the Nuget packet DotNetStandard 2.0 for this project.
So I tried VS2019 with framework 4.6.1, same sample program with the same output results.

I then simultaneously captured your capture sample program (vs2015 version) along w/ wireshark to compare the results.
I compared the two (@same timetag) and found the difference seems to be that SharpPcap doesn't recognize Encapsulation type Null/Loopback. I attached pics of the comparison of the two.

![WiresharkCapture](https://user-images.githubusercontent.com/81334566/113346665-c3894f80-92f9-11eb-8bc4-6fdda8d02e63.JPG)
![SharPcapCapture](https://user-images.githubusercontent.com/81334566/113346677-c71cd680-92f9-11eb-8b14-f3541a844c8d.JPG)

Thanks for your help!
Terry

**Comment by Terry550** [comment]
I noticed Closed Issue #17 seems to talk about the same issue.
I'm using Nuget release SharpPcap v5.4 and PacketDotNet v1.2, Did the fix get rolled in a later version?
Did it get fixed for both Linux and windows?
Thanks,
Terry

**Comment by chmorgan** [comment]
@Terry550 what capture library are you using? npcap 1.2? I'm asking because looking at your code there shouldn't be any modification to the packets by SharpPcap or npcap. So maybe Wireshark is writing to it differently?

**Comment by Terry550** [comment]
Hi chmorgan,
Thanks for getting back with me. 
I'm using NPcap v1.20 and the file versions found in the System Npcap folder are:
wpcap.dll (1.10.1 - PRE-GIT)
packet.dll (5.1.20.305)

In your example I had to modify the following line because it would compile for me:
device.Open(mode: DeviceModes.Promiscuous | DeviceModes.DataTransferUdp | DeviceModes.NoCaptureLocal, read_timeout: readTimeoutMilliseconds);
to
device.Open(mode: DeviceMode.Promiscuous, read_timeout: readTimeoutMilliseconds);
Maybe that will be a hint.

Have you had a chance to run my solution to see if it gives you the same pcap output?
Thanks,
Terry

**Comment by Terry550** [comment]
I'm wondering if you have a newer version of libpcap since it says "PRE-GIT" when I look at the properties.
Also, maybe the newer version allows this device.Open method:
device.Open(mode: DeviceModes.Promiscuous | DeviceModes.DataTransferUdp | DeviceModes.NoCaptureLocal, read_timeout: readTimeoutMilliseconds);

**Comment by kayoub5** [comment]
@Terry550 how did you create the pcap writer, and did you create it before or after opening the device?

**Comment by Terry550** [comment]
I created the capturefileWriterdevice as a static and opened the writer
after opening the device.
Today I ran the NPCAP example called savedump.c on the loop back adapter
and the capture looked good in wireshark but still can’t get the Sharp Pcap
to work.
Thanks
Terry

**Comment by kayoub5** [comment]
that would be the issue, you did not provide the link layer argument to the writer, so it's using ethernet link layer by default

**Comment by Terry550** [comment]
Should the link layer type be null?

**Comment by kayoub5** [comment]
The link layer is device dependent, see LinkLayer property of the device

**Comment by Terry550** [comment]
Yea! That worked. So if I want to capture loopback traffic and non-loopback
traffic to the same Pcap File is that possible using only one capture
writer. Or can I use multiple capture writers writing to the same Pcap file?

**Comment by kayoub5** [comment]
In a normal pcap file, you can't mix multiple link layers, it's a file format restriction, not a writer restriction. 
however, pcapng files can do it

**Comment by chmorgan** [comment]
Doesn’t libpcap support pcapng format now? Is there an API flag that lets
the user select which format is to be used? I can’t remember seeing any.

**Comment by kayoub5** [comment]
Ironically, libpcap support for PCAPNG is extremely limited:
* It only supports reading PCAPNG
* It only supports reading a PCAPNG that contains one link type, see https://github.com/the-tcpdump-group/libpcap/blob/bcca74d2713dc9c0a27992102c469f77bdd8dd1f/sf-pcapng.c#L1237

**Comment by Terry550** [comment]
Thank you guys for your EXPERTISE!! And solving my issues!

**Comment by kayoub5** [comment]
@chmorgan both this issue and #174 originate from the fact that the writer does not check that the packet link layer and the file link layer matches, maybe we should add a check and make it throw to avoid any confusion?

**Comment by chmorgan** [comment]
Yeah that does seem like a good idea. There doesn’t seem to be a better way
to do without checking each packet...
